lib/hal/xbox.c

```diff
@@ -9,51 +9,78 @@
 
 #define KernelMode 0
 
+#define LaunchDataPageSize 0x1000
+
 void XReboot()
 {
-	HalReturnToFirmware(HalRebootRoutine);
+    HalReturnToFirmware(HalRebootRoutine);
+}
+
+int XGetLaunchInfo(unsigned long *launchDataType, const unsigned char **launchData)
+{
+    *launchDataType = LDT_NONE;
+    *launchData = NULL;
+
+    if (LaunchDataPage == NULL) {
+        LaunchDataPage = MmAllocateContiguousMemory(LaunchDataPageSize);
+        if (LaunchDataPage == NULL) {
+            return -1;
+        }
+        memset(LaunchDataPage, 0, LaunchDataPageSize);
+        LaunchDataPage->Header.dwLaunchDataType = LDT_FROM_DASHBOARD;
+    }
+
+    *launchDataType = LaunchDataPage->Header.dwLaunchDataType;
+    *launchData = LaunchDataPage->LaunchData;
+    return 0;
 }
 
-/**
- * Launches an XBE.  Examples of xbePath might be:
- *   c:\\blah.xbe
- *   c:/foo/bar.xbe
- * If the XBE is able to be launched, this method will
- * not return.  If there is a problem, then it return.
- */
-void XLaunchXBE(char *xbePath)
+void XLaunchXBE(const char *xbePath)
 {
-#if 0
-	struct stat statbuf;
-	if (stat(xbePath, &statbuf) < 0)
-		return;
-#endif
+    XLaunchXBEEx(xbePath, NULL);
+}
 
-    if (LaunchDataPage == NULL)
+void XLaunchXBEEx(const char *xbePath, const void *launchData)
+{
+    if (LaunchDataPage == NULL) {
         LaunchDataPage = MmAllocateContiguousMemory(0x1000);
+        if (LaunchDataPage == NULL) {
+            return;
+        }
+    }
+
+    // For ease of debugging.
+    PLAUNCH_DATA_PAGE launchDataPage = LaunchDataPage;
+
+    MmPersistContiguousMemory(launchDataPage, LaunchDataPageSize, TRUE);
+    memset((void*)launchDataPage, 0, LaunchDataPageSize);
 
-    if (LaunchDataPage == NULL)
-		return;
+    launchDataPage->Header.dwLaunchDataType = LDT_TITLE;
+    launchDataPage->Header.dwTitleId = CURRENT_XBE_HEADER->CertificateHeader->TitleID;
+    launchDataPage->Header.dwFlags = 0x0000;
 
-    MmPersistContiguousMemory(LaunchDataPage, 0x1000, TRUE);
+    if (!xbePath) {
+        launchDataPage->Header.dwLaunchDataType = LDT_LAUNCH_DASHBOARD;
+    } else {
+        XConvertDOSFilenameToXBOX(xbePath, launchDataPage->Header.szLaunchPath);
 
-	memset((void*)LaunchDataPage, 0, 0x1000);
+        // one last thing... xbePath now looks like:
+        //   \Device\Harddisk0\Partition2\blah\doom.xbe
+        // but it has to look like:
+        //   \Device\Harddisk0\Partition2\blah;doom.xbe
+        char *lastSlash = strrchr(launchDataPage->Header.szLaunchPath, '\\');
+        if (!lastSlash) {
+            // if we couldn't find a trailing slash, the conversion to
+            // the xbox path mustn't have worked, so we will return
+            return;
```

Comment:
nit: What should happen to the `LaunchDataPage` here? Should the allocation be undone?